### ISSUE TITLE
Cannot push down count(distinct) queries on distribution column / group by distribution column for hash partitioned tables

### ISSUE THREAD

**Issue by samay-sharma** [citusdata/citus#311]
When a table is distributed by hash, we know that we don't have overlapping shards. Therefore we should be able to push down the `count(distinct)` if it is on the distribution column or if the query has a group by on the distribution column. We do so for range partitioned tables. However, we don't do that for hash partitioned tables.

When `lineitem` is distributed by `l_orderkey`, the following query fails:

```sql
select count(distinct l_orderkey) from lineitem group by l_orderkey;
ERROR:  cannot compute aggregate (distinct)
DETAIL:  table partitioning is unsuitable for aggregate (distinct)
HINT:  You can load the hll extension from contrib packages and enable distinct approximations.
```

**Comment by ozgune** [comment]
@samay-sharma, could you comment on this issue with your understanding?

@lithp, was this query returning incorrect results? Is the correctness here related to our join method (broadcast, co-located, or repartitioned) at all?

**Comment by samay-sharma** [comment]
@lithp : I'll look into this tomorrow and send a reply.

**Comment by lithp** [comment]
> I also fixed what I think was a bug by disallowing any GROUP BY which isn't on the partition column.

After thinking about it some more, you can ignore everything after that line. It's fine to GROUP BY another column as long as you're count distinct'ing the partition column. I updated the PR and left that query as-is.

The PR now allows count distinct on hash-partitioned tables but disallows joins involving hash-partitioned tables. I couldn't find a case in which joins didn't work, but also couldn't prove they would give correct results under every type of join we have, so to be safe disallowed them.  I think range partitioning joins aren't safe in the general-case either but they're currently being used in production.

**Comment by samay-sharma** [comment]
Hey @lithp : Looking at the code for range partitioned tables, I think the only join which will go through our checks is if the count(distinct) aggregate is on the large table's partition column (or group by is on partition column) and the other table is a one shard table.

This is because the matching with the partition column is done on the basis of `VarNo` and `VarAttNo` for all tables in the `tableNodeList`. And I think `VarNo` and `VarAttNo` won't be same for columns of different tables.

So, if there are two co-located tables (for eg.), then the `tablePartitionColumn` won't match for both tables and hence one of them would result in `tableDistinctSupported` as being `false`. This would lead to us error'ing out for that.

Could you check if that is indeed the case by creating two colocated range partitioned tables and executing count(distinct) with joins?

And if we produce correct results for that in the range case, I think we should also produce correct results in the hash case.

I agree that this checking is a bit implicit, but its worth verifying that this check disallows joins for which we could produce incorrect results even in the range case.

**Comment by marcocitus** [comment]
I spent some time looking into this so make me sleep more comfortably, and it doesn't seem like there are any cases which cause incorrect results, under the assumption that shards are strictly disjoint, which we generally assume for range- and hash-partitioned tables.

I created two distributed tables: visits range-partitioned by customer_id with 3 shards, and pages range-partitioned by page_id with 2 shards. I was mostly concerned with re-partition joins and broadcast joins, but re-partition joins are already allowed. In a broadcast join, each shard of visits is joined with both shards of pages and the results of matching group by combinations are summed on the master. There are cases where this could lead to double counting, which the check must prevent.

I verified the correct behaviour in the following cases, which would should probably add to our regression tests:

-- Group by all partition keys. Each join deals with a disjoint set of (customer_id, page_id) combinations, and therefore no summing occurs.
SELECT customer_id, page_id, count(distinct url) FROM visits, pages WHERE visits.site_id = pages.site_id GROUP BY customer_id, page_id;

-- Count unique partition key values. Each join deals with a disjoint set of (customer_id, page_id) combinations, and therefore it is safe to sum the distinct page_id counts without double counting.
SELECT customer_id, count(distinct page_id) FROM visits, pages WHERE visits.site_id = pages.site_id GROUP BY customer_id;

-- Allowed if pages has a single shard, in which case each join deals with a disjoint set of customer_id's, disallowed otherwise.
SELECT customer_id, count(distinct url) FROM visits, pages WHERE visits.site_id = pages.site_id GROUP BY customer_id;

-- Allowed if pages has a single shard, in which case each join deals with a disjoint set of customer_id's, disallowed otherwise.
SELECT customer_id, count(distinct page_id), count(distinct url) FROM visits, pages WHERE visits.site_id = pages.site_id GROUP BY customer_id;

The check would not work for re-partition queries, but we already disallow count distinct with re-partitioning.

It should be safe to allow count distinct push-down for joins of hash-partitioned tables, including a mixture of hash- and range-partitioned tables, just not for append-partition tables because of the potential overlap.
